File: program/program/move-all-zeroes-to-end-of-array/move_all_zeroes_to_end_of_array.r
moveZeroesToEnd <- function(arr) {
  nonZeroIndex <- 1

  for (i in 1:length(arr)) {
    if (arr[i] != 0) {
      arr[nonZeroIndex] <- arr[i]
      nonZeroIndex <- nonZeroIndex + 1
    }
  }

  for (i in nonZeroIndex:length(arr)) {
    arr[i] <- 0
  }

  return(arr)
}


arr1 <- c(1, 2, 0, 4, 3, 0, 5, 0)
result1 <- moveZeroesToEnd(arr1)
cat("Input  : arr[] =", arr1, "\n")
cat("Output : arr[] =", result1, "\n")

arr2 <- c(1, 2, 0, 0, 0, 3, 6)
result2 <- moveZeroesToEnd(arr2)
cat("Input  : arr[] =", arr2, "\n")
cat("Output : arr[] =", result2, "\n")

File: program/move-all-zeroes-to-end-of-array/MoveAllZeroesToEndOfArray.js
function moveZeroesToEnd(arr) {
    let count = 0; // Count of non-zero elements
    
    // Traverse the array. If element encountered is non-zero, then replace the element at index 'count' with this element
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] !== 0) {
            arr[count++] = arr[i];
        }
    }
    
    // Now all non-zero elements have been shifted to front and 'count' is set as index of first 0. Make all elements 0 from count to end
    while (count < arr.length) {
        arr[count++] = 0;
    }
    
    return arr;
}

// Test cases
let arr1 = [1, 2, 0, 4, 3, 0, 5, 0];
console.log(moveZeroesToEnd(arr1)); // Output: [1, 2, 4, 3, 5, 0, 0, 0]

let arr2 = [1, 2, 0, 0, 0, 3, 6];
console.log(moveZeroesToEnd(arr2)); // Output: [1, 2, 3, 6, 0, 0, 0]
